### DISCUSSION TITLE
Is `toNumbers` supposed to mutate data?

### DISCUSSION THREAD

**Discussion by scpedicini** [nlp-compromise/nlp_compromise#1112] [Q&A]
I think I'm doing something wrong with the newer `compromise/three`. `termList`/`out`/etc. works great, but if I use them in conjunction with `toNumbers` the data seems to mutate in place.

Trying to do this:

```ts
const text_1 = "There are twenty-four apples and 12000 oranges on the table.";
const nat = nlp(text_1);

const number_of_numerical_values = nat.values().length;
// shows a length of two which is what we would expect so far so good

const termList_1 = nat.values(0).toNumber().termList();
console.log(termList_1);
// termList_1 has converted the number "twenty-four" to "24" but also has
// `"apples", "and", "12000", "oranges", "on", "the", "table"` in array

// rerun exact same command
const termList_2 = nat.values(0).toNumber().termList();
console.log(termList_2);
// termList_2 now has the expected array of just "3" 
// converted to numerical
```

I basically want to go through each numerically recognized "group" hence using the `.values()` command and use the `toNumbers()` command to reduce them down to the numerical number.

Picture to help indicate the issue:

![CleanShot 2024-06-03 at 21 55 07@2x](https://github.com/spencermountain/compromise/assets/2040540/225b6821-9033-49d5-a1d6-9d2ccba89d59)

**Comment by spencermountain** [comment]
oh weird - ya, i see what you mean.

I think it may (first) be a bug with `.termList()`. I would use `.json()` and flatten it:
```js
nat.values(0).toNumber().json(); // seems to work
```
if I can remember, I think `.toNumbers()` may return `.all()` because it was returning a different number of terms - "five hundred and fifty seven" (5 terms)-> "557" (1 term). Although I'd have to check that. That may be why it is inconsistent - the second time it is not making a transformation, so it just passes through. Not great.

oh boy. I can checkout why things are goofed up. May be two separate things
cheers

**Reply by spencermountain** [reply]
made an issue for this here https://github.com/spencermountain/compromise/issues/1113

**Reply by scpedicini** [reply]
Hmmm that one failed for me too - using json(), out(), termList() all seem to result in an "in-place" transformation.

![CleanShot 2024-06-04 at 12 32 15@2x](https://github.com/spencermountain/compromise/assets/2040540/5dffe091-9b02-490d-a8c9-a50d6df7e27d)


I can reduce at the beginning to json(), e.g.

```ts
const json = nat.values(0).json()
const json_2 = nat.values(1).json()
```

But unfortunately now I don't get to take advantage of the nice feature of `toNumber()` which reduces numerical words into numbers (e.g. three => 3).

![CleanShot 2024-06-04 at 12 34 22@2x](https://github.com/spencermountain/compromise/assets/2040540/fe55cf6d-7317-41f5-8473-e10f289a650d)

**Reply by scpedicini** [reply]
@spencermountain I think I'm going to temporarily work around it by instantiating another copy of `nlp` for each value, e.g.

```ts
import nlp from "compromise/three";

const nat = nlp("There were twenty-four apples and 12000 oranges on the table.");
// @ts-ignore
for(let i = 0; i < nat.values().length; i++) {
    // @ts-ignore
    const nat_2 = nlp(nat.values(i).out()).values(0).toNumber().termList();
    console.log(nat_2);
}
```

Does that seem like an acceptable workaround?

Also just a random aside, are there some typedefs I can use so I don't have to continually use these `ts-ignore` directives? EDIT: Actually on closer inspection I think its just missing the `values()` typedef:

```
TS2339: Property values does not exist on type Three
```

Thanks!

**Reply by spencermountain** [reply]
hi Shaun - ya, if you wanted to get all the parsed numbers from the document, i would do
```js
nat.values().json() // and loop through that
```
if you want to transform them in the document, i would do:
```js
nat.values().toNumber() // drop this, because of the mentioned bug
nat.values().forEach(m=>{
  //do whatever with each transformed number, in sequence
})
```
typescript should work! Please file an issue if you believe the types are configured in the wrong way.
cheers
